dataplicity/m2m/commandservice.py

```diff
@@ -38,54 +38,83 @@ def __init__(self, channel, command):
     def __repr__(self):
         return self._repr
 
+    @classmethod
+    def send_error(cls, channel, status, msg, **extra):
+        """Send a control packet with an error"""
+        error = {
+            "service":"run-command",
+            "type": "error",
+            "status": status,
+            "msg": msg
+        }
+        error.update(extra)
+        channel.send_control(error)
+
     def run_service(self, channel, command):
         """Run the thread and log exceptions."""
         try:
             self._run_service(channel, command)
         except Exception:
             log.exception("error running %r", self)
+            self.send_error(channel, "error", "error running command")
 
     def _run_service(self, channel, command):
         """Run command and send stdout over m2m."""
         log.debug("%r started", self)
-        try:
-            process = subprocess.Popen(
-                command,
-                stdout=subprocess.PIPE,
-                shell=True
-            )
-        except OSError as error:
-            log.warning('%r command failed; %s', self, error)
-            return
-
+        process = subprocess.Popen(
+            command,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            shell=True
+        )
         bytes_sent = 0
-        fh = process.stdout.fileno()
+        stdout_fh = process.stdout.fileno()
+        stderr_fh = process.stderr.fileno()
         try:
             while True:
                 if channel.is_closed:
                     log.debug("%r channel closed", self)
                     break
-                readable, _, _ = select.select([fh], [], [], 0.5)
-                if readable:
-                    chunk = os.read(fh, self.CHUNK_SIZE)
+                readable, _, _ = select.select(
+                    [stdout_fh, stderr_fh], [], [], 0.5
+                )
+
+                if stdout_fh in readable:
```

Comment:
stdout get sent over m2m, stderr gets written to logs. This is so that the remote end doesn't interpret errors as valid instructions.